src/libsystemd/sd-device/sd-device.c

```diff
@@ -1457,15 +1457,25 @@ static int device_properties_prepare(sd_device *device) {
                 return r;
 
         if (device->property_devlinks_outdated) {
-                char *devlinks = NULL;
+                _cleanup_free_ char *devlinks = NULL;
+                size_t devlinks_allocated = 0, devlinks_len = 0;
+                char *e;
                 const char *devlink;
 
                 devlink = sd_device_get_devlink_first(device);
-                if (devlink)
-                        devlinks = strdupa(devlink);
-
-                while ((devlink = sd_device_get_devlink_next(device)))
-                        devlinks = strjoina(devlinks, " ", devlink);
+                if (devlink) {
+                        if (!GREEDY_REALLOC(devlinks, devlinks_allocated, strlen(devlink) + 1))
+                                return -ENOMEM;
+                        e = stpcpy(devlinks, devlink);
+                        devlinks_len = e - devlinks;
+
+                        while ((devlink = sd_device_get_devlink_next(device))) {
+                                if (!GREEDY_REALLOC(devlinks, devlinks_allocated, devlinks_len + strlen(devlink) + 2))
+                                        return -ENOMEM;
+                                e = stpcpy(stpcpy(devlinks + devlinks_len, " "), devlink);
+                                devlinks_len = e - devlinks;
+                        }
```

Comment:
Seems fine, this would make it similar to the `tags` below, except with the special-casing of not prepending a space with the first entry. I'll change it accordingly tomorrow morning.